riyaz/cli.py

```diff
@@ -0,0 +1,141 @@
+"""
+Documentation for Riyaz CLI
+
+The CLI can run these commands:
+    - `riyaz serve`: Start a local instance of Riyaz. Useful for local course
+                     development
+    - `riyaz new`: Start a new Riyaz course with the default template
+    - `riyaz push`: Push the Riyaz course in current directory to a remote
+                    server
+    - `riyaz pull`: Pull a Riyaz course from a remote server
+"""
+
+import os
+import re
+from pathlib import Path
+from typing import List, IO, Optional, Union
+
+import click
+import frontmatter
+import yaml
+from pydantic import BaseModel, HttpUrl, validator
+from pydantic.types import FilePath
+
+
+class ChapterConfig(BaseModel):
+    name: str
+    title: str
+    lessons: List[FilePath]
+
+
+class Config(BaseModel):
+    name: str
+    title: str
+    short_description: Optional[str]
+    description: str
+    authors: List[str]
+    outline: List[ChapterConfig]
+
+    @validator("authors", each_item=True)
+    def check_author_file_exists(cls, v):
+        author_file = get_author_file_path(v)
+        if not author_file.exists():
+            raise ValueError(f"author file {author_file} does not exist")
+
+        return v
+
+
+class Lesson(BaseModel):
+    name: str
+    title: str
+    path: FilePath
+
+    @classmethod
+    def from_path(cls, path: FilePath) -> "Lesson":
```

Comment:
You can use forward references like `-> Lesson` when defining the `Lesson` class if you add the following import at the beginning on the file.

```
from __future__ import annotations
```

Reply:
One small suggestion. You seem to have two different sets of pydanantic models - one for the format on disk and another for that we use in the db/api. It is confusing to see them in the same place, there is also conflict of names so you had to use `Config` for `course.yml` etc.  

Here is a suggestion to resolve that ambiguity. Use  a prefix like `Disk` for all the models dealing with format on disk - like `DiskCourse`, `DiskChapter`, `DiskLesson` etc. Move the models that the system actually uses to a different file.

How about:
- rename `models.py` to `doctypes.py`
- add a `models.py` with pydantic models that we use in riyaz
- add a `disk.py` with pydantic models that we use for on disk format and othe utility functions

Reply:
Also for the testing the bad format, it may be too cumbursome to create files for every possible failure. It may be good idea to have one sample course in the right format and specify defects in code and verify if the system catches that error. 

For this, instead of depending on files being present in a particular location, you can use `tmp_path` fixture to create a new course structure by copying the sample course data and add defects as needed and test.

Something like:

```
class BaseDefectsTest:
    ....
    
class TestInvalidAuthors(BaseDefectsTest):
     def add_defects(self):
          with self.patch("course.yaml") as d:
              d['authors'] = "alice"

class TestInvalidChapters(BaseDefectsTest):
     def add_defects(self):
          with self.patch("course.yaml") as d:
              del d['outline'][0]['title]
```

Reply:
I have made edits for the first two review comments.
It would be very helpful to include the third too, but I have kept it for later. I will work on it as I write more tests.